pkg/upgrade/migrate.go

```diff
@@ -338,23 +338,20 @@ func Migrate(ctxt context.Context, c client.Interface, nodename string) error {
 		log.Info("successfully released lock on host-local backend!")
 	}
 
-	// Re-enable datastoreReady if this is the last node to update. We know this is the last
-	// node to update if the number of UpdatedNumberScheduled is MaxUnavailable less than the
-	// DesiredNumberScheduled.
-	calicoDS, err := k8sClient.AppsV1().DaemonSets("kube-system").Get(context.Background(), "calico-node", metav1.GetOptions{})
-	if err != nil {
-		log.WithError(err).Error("failed to retrieve Calico daemonset to check if datastore readiness should be set to true")
-	} else if calicoDS.Status.UpdatedNumberScheduled == calicoDS.Status.DesiredNumberScheduled-calicoDS.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable.IntVal {
-		log.Info("setting Calico datastore readiness to true...")
-		t := true
-		clusterInfo.Spec.DatastoreReady = &t
-		if _, err = c.ClusterInformation().Update(ctxt, clusterInfo, options.SetOptions{}); err != nil {
-			return fmt.Errorf("failed to re-enable cluster: %s", err)
-		}
-		log.Info("successfully set Calico datastore readiness to true!")
-	} else {
-		log.Info("not setting datastore readiness, rolling update is still going")
+	// Always re-enable datastoreReady. This leaves a very small chance that an address may be assigned by calico-ipam
+	// that is already in use by host-local. Any pods stuck in this situation will be fixed once they are restarted.
+	// TODO: Re-enable datastoreReady if all of the nodes have been picked up for the rolling update.
+	// This logic has not been used because of the potential length of time it would bring normal operation
+	// of the cluster down for by locking the datastore until the update has been applied to all nodes.
+	// calicoDS, err := k8sClient.AppsV1().DaemonSets("kube-system").Get(context.Background(), "calico-node", metav1.GetOptions{})
+	// if calicoDS.Status.UpdatedNumberScheduled == calicoDS.Status.DesiredNumberScheduled {
+	log.Info("setting Calico datastore readiness to true...")
+	t := true
+	clusterInfo.Spec.DatastoreReady = &t
+	if _, err = c.ClusterInformation().Update(ctxt, clusterInfo, options.SetOptions{}); err != nil {
```

Comment:
Did you consider using a patch operation here instead of an update? This update can fail if another Kubernetes client changed the _ClusterInformation_ object since we acquired our copy of it (in the `clusterInfo` variable here). Ideally, we'd have our setting this value back to true succeed regardless of intervening changes from others. Patching would allow us that unconditional success.

Reply:
That's a good point and probably a cleaner way to do this than we have now. Right now, if we hit the error updating since the `ResourceVersion` is no longer the most current in the case you've outlined, we'll end up retrying the entire migration again until it works and passes (which isn't nearly as nice and clean as you've outlined with `Patch` https://github.com/projectcalico/cni-plugin/blob/master/pkg/ipamplugin/ipam_plugin.go#L97-L104). So I would expect us to work in that case if I'm understanding it correctly, it's just not nearly as nice or efficient as it could be. 

Also, it appears we haven't implemented `Patch` yet for any of our resource APIs (`ClusterInfo` client for reference https://github.com/projectcalico/libcalico-go/blob/master/lib/clientv3/clusterinfo.go) so we would need to update that first before we could use it. Definitely something we should be doing in the future, but for now, in order to get this out ASAP, we'll be skipping it.